### PR TITLE
EES-2853 & EES-2854 table size custom error and ga event

### PR DESCRIPTION
**Pull request by amyb-hiveit**
Adds custom error handling for the table size error to:
- show a button to download the subject file in the error message (in the admin this button is disabled / not shown as appropriate)
- log a GA event

**Public:**
tablesizeerror-public

**Admin - table tool preview and pre-release:**
admin

**Admin - data blocks**
tablesizeerror-datablock

### CODE REVIEW

**Comment by benoutram** [comment]
I'm not sure if 'Select different filters or download the subject data' is the right wording for the Admin app as the analysts are trying to build data blocks so they wouldn't really need to download data that they are working with but they do need to know that they need to change the combination of filters.

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  const ref = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    if (focus && ref.current) {
+      ref.current.focus();
+    }
+  }, [focus]);
+
+  return (
+    <div
```
Comment (ntsim) [inline src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx:35]: Seems like we could extract out a `BaseErrorSummary` (exported from the same file as `ErrorSummary`). It would be a dumb component that takes basic things like the `children`, `tiitle` and `id` and would be used by `ErrorSummary` itself.

We would then be able to re-use the `BaseErrorSummary` here instead of having to build another version of the component here.

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    >
+      <h2 className="govuk-error-summary__title" id="tableSizeError">
+        There is a problem
+      </h2>
+
+      <ErrorPrefixPageTitle />
+      <div className="govuk-error-summary__body">
+        <p>
+          A table cannot be returned as the filters chosen can exceed the
+          maximum allowable table size.
```
Comment (ntsim) [inline src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx:51]: This error message is a little weird - would suggest rewording it to 'Could not create table as the filters chosen may exceed the maximum allowed table size'

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                className="govuk-!-margin-bottom-0"
+                onClick={() => handleDownloadFile(subject.file.id)}
+              >
+                {`Download ${subject.name} (${subject.file.extension}, ${subject.file.size})`}
+              </Button>
+            ) : (
+              <Button className="govuk-!-margin-bottom-0" disabled>
+                Download subject file (available when the release is published)
+              </Button>
+            )}
```
Comment (ntsim) [inline src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx:67]: Bit messy to use the ternary over the entire button as we end up having to declare it twice. I'm also not sure why the messages so different as well?

We could be a bit more declarative with something like:

```tsx
<Button
  className="govuk-!-margin-bottom-0"
  disabled={!releaseId}
  onClick={() => handleDownloadFile(subject.file.id)}
>
  {`Download ${subject.name} (${releaseId 
    ? `${subject.file.extension}, ${subject.file.size}` 
    : 'available when the release is published'})`}
</
```

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  focus = false,
+  releaseId,
+  showDownloadOption = true,
+  subject,
+}: Props) => {
+  const handleDownloadFile = async (fileId: string) => {
+    if (releaseId) {
+      await downloadService.downloadFiles(releaseId, [fileId]);
+    }
+  };
```
Comment (ntsim) [inline src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx:24]: This is pretty simple, so I'd suggest we just inline this into the `onClick` prop itself.

File: src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx [hunk 5]
```diff
@@ -105,10 +115,33 @@ const FiltersForm = (props: Props & InjectedWizardProps) => {
...
+    } catch (error) {
+      if (isServerValidationError(error) && error.response?.data) {
+        const errorMessages = Object.entries(error.response.data.errors).reduce(
+          (acc, [, messages]) => {
+            messages.forEach(message => acc.push(message));
+            return acc;
+          },
+          [] as string[],
+        );
+        if (errorMessages.includes('QUERY_EXCEEDS_MAX_ALLOWABLE_TABLE_SIZE')) {
```
Comment (ntsim) [inline src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx:133]: We seem to be doing a lot of work here when we it can probably boil it down to something like this:

```ts
if (isServerValidationError(error)) {
  const errors = Object.values(error.response?.data.errors);
  
  if (errors.includes('QUERY_EXCEEDS_MAX_ALLOWABLE_TABLE_SIZE')) {
    // do stuff... 
```

File: src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx [hunk 6]
```diff
@@ -36,7 +42,14 @@ interface Props {
     indicators: string[];
     filters: string[];
   };
+  selectedPublication?: SelectedPublication;
+  subject: Subject;
   subjectMeta: SubjectMeta;
+  tableSizeErrorDownloadAvailable?: boolean;
+  tableSizeErrorLogEvent?: (
```
Comment (ntsim) [inline src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx:49]: This is an event handler so we should name this according to those conventions i.e. `onTableSizeError`

File: src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx [hunk 7]
```diff
@@ -36,7 +42,14 @@ interface Props {
     indicators: string[];
     filters: string[];
   };
+  selectedPublication?: SelectedPublication;
+  subject: Subject;
   subjectMeta: SubjectMeta;
+  tableSizeErrorDownloadAvailable?: boolean;
```
Comment (ntsim) [inline src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx:48]: This is a bit of a mouthful with the 'available' - could simplify this to `showTableSizeErrorDownload`

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+interface Props {
+  focus: boolean;
+  releaseId?: string;
+  showDownloadOption?: boolean;
+  subject: Subject;
+}
+
+const TableSizeError = ({
+  focus = false,
```
Comment (ntsim) [inline src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx:15]: Doesn't seem like we really need this to be configurable

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableToolWizard.tsx [hunk 9]
```diff
@@ -68,11 +68,16 @@ export interface TableToolWizardProps {
   initialState?: Partial<InitialTableToolState>;
   hidePublicationSelectionStage?: boolean;
   finalStep?: (props: FinalStepRenderProps) => ReactElement;
+  loadingFastTrack?: boolean;
   renderFeaturedTable?: (featuredTable: FeaturedTable) => ReactNode;
   scrollOnMount?: boolean;
+  tableSizeErrorDownloadAvailable?: boolean;
+  tableSizeErrorLogEvent?: (
```
Comment (ntsim) [inline src/explore-education-statistics-common/src/modules/table-tool/components/TableToolWizard.tsx:75]: Would make name changes here based on comments in `FiltersForm` too

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    if (focus && ref.current) {
+      ref.current.focus();
+    }
+  }, [focus]);
+
+  return (
+    <div
+      aria-labelledby="tableSizeError"
+      className="govuk-error-summary"
+      id="filtersForm-tableSizeError"
```
Comment (ntsim) [inline src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx:38]: We should pass the id in as a prop from `FiltersForm` and use the `formId` variable like so:

```tsx
<TableSizeError
  id={`${formId}-tableSizeError`}
>
```

**Review by ntsim — CHANGES_REQUESTED** [review]

**Comment by duncan-at-hiveit** [comment]
Hey @amyb-hiveit - I noticed in the screenshot under "Admin - table tool preview and pre-release", there was no Subject name displayed.  Is this intentional, or do we have a missing Subject name coming back from the API call in this instance?

**Comment by amyb-hiveit** [comment]
> Hey @amyb-hiveit - I noticed in the screenshot under "Admin - table tool preview and pre-release", there was no Subject name displayed. Is this intentional, or do we have a missing Subject name coming back from the API call in this instance?

Well spotted! That's just me adding the wrong screenshot. Here's what it actually looks like:
admin

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx [hunk 11]
```diff
@@ -1,76 +1,57 @@
...
+            disabled={!releaseId}
+            onClick={async () => {
+              if (releaseId) {
+                await downloadService.downloadFiles(releaseId, [
+                  subject.file.id,
+                ]);
+              }
+            }}
+          >
+            {`Download ${subject.name} (${subject.file.extension}, ${subject.file.size})`}
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx:47]: As I mentioned as an overall PR comment, the screenshots have the Preview mode not showing the Subject name.  Perhaps something to quickly look at?

File: src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx [hunk 12]
```diff
@@ -123,16 +120,10 @@ const FiltersForm = (props: Props & InjectedWizardProps) => {
       goToNextStep();
     } catch (error) {
       if (isServerValidationError(error) && error.response?.data) {
```
[the remark this answers is not in the document]
Comment (amyb-hiveit) [reply src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx]: Good idea, I've combined it with isServerValidationError.

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx (hunk 11 above)
Comment (amyb-hiveit) [reply src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx:47]: It was the screenshot that was wrong not the code, doh! I've updated it now.

**Review by duncan-at-hiveit — APPROVED** [review]

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx (hunk 1 above)
Comment (duncan-at-hiveit) [reply src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx]: Happy that this has suggestion has been addressed

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx (hunk 2 above)
Comment (duncan-at-hiveit) [reply src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx]: Happy that this has suggestion has been addressed

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx (hunk 3 above)
Comment (duncan-at-hiveit) [reply src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx]: Happy that this has been addressed :)

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx (hunk 4 above)
Comment (duncan-at-hiveit) [reply src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx]: Happy that this has been addressed :)

File: src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx (hunk 5 above)
Comment (duncan-at-hiveit) [reply src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx]: Happy that this has been addressed :)

File: src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx (hunk 6 above)
Comment (duncan-at-hiveit) [reply src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx]: Happy that this has been addressed :)

File: src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx (hunk 7 above)
Comment (duncan-at-hiveit) [reply src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx]: Happy that this has been addressed :)

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableToolWizard.tsx (hunk 9 above)
Comment (duncan-at-hiveit) [reply src/explore-education-statistics-common/src/modules/table-tool/components/TableToolWizard.tsx]: Happy that this has been addressed :)

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx (hunk 10 above)
Comment (duncan-at-hiveit) [reply src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx]: Happy that this has been addressed :)

File: src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply src/explore-education-statistics-common/src/modules/table-tool/components/FiltersForm.tsx]: Happy that this has been addressed :)

File: src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx (hunk 11 above)
Comment (duncan-at-hiveit) [reply src/explore-education-statistics-common/src/modules/table-tool/components/TableSizeError.tsx:47]: Happy that this has been addressed :)
